### DISCUSSION TITLE
The state machine in C# with async/await

### DISCUSSION THREAD

**Discussion by giscus** [linkdotnet/Blog.Discussions#113] [General]
# The state machine in C# with async/await

You often here that the async/await keywords leads to a state machine. But what does that mean? Let's discuss this with a simple example.

https://steven-giesel.com/blogPost/720a48fd-0abe-4c32-83ac-26926d501895

**Comment by kapsiR** [comment]
Worth noting that a tool [like SharpLab](https://sharplab.io/#v2:CYLg1APgAgTAjAWAFBQAwAIpwKwG5nJQDMmM6AwusgN7Lr2ZwBsmAHJiwCoAWATgPYB3AKIAPAMYBTAA4AXAJb8AdgEEAzgE8l4gBQBKOg1pIGpzAE4OAOgCa8yQBtg+/CbP1ZfIeiWTB6MSk5RSUdACIACUkNML1XUwBfZASgA=) is incredibly useful to understand it.

Regarding exceptions: It is important to understand, that the exceptions are rethrown at the first `await`. (eliding it will ignore all exceptions that are stored in the task)

**Reply by linkdotnet** [reply]
Probably will add a part to that - if the exception happens in the synchronous part, the exception will be thrown.

**Comment by sksk571** [comment]
> if the exception happens in the synchronous part, the exception will be thrown.

No, it won't. This code doesn't throw exception:

```
var t = ThrowExceptionAsync();
// No exception;

static async Task ThrowExceptionAsync()
{
    throw new Exception("Hey");
    await Task.Yield();
}
```

**Reply by linkdotnet** [reply]
You are correct - corrected! Thanks

**Comment by toupswork** [comment]
I was looking into why the lock keyword can't be used in async methods (although it can be in .net 9, it can't cross await boundaries) and I learned something new. I didn't realize that after the await, the thread that resumes after await has a new and different stack frame. It makes sense considering that async has no thread affinity across await, because the thread doesn't get blocked and can go off and do other work. But then I asked myself, well how does the resuming thread know which method to return to when the async method ends. I was sort of thinking that the resuming thread assumes the call stack of the previous thread, which is analogous to green threads, which of course .NET does not have. When a function normally ends, it has the return address that was pushed onto the stack, but if the resuming thread has a different call stack, then how exactly does it not where to return to? So internally the state machine uses a delegate which points the continuation addresss. Did you already know this?

**Reply by linkdotnet** [reply]
Well, to some extent. I was aware that Threads (and, to some extent, Tasks, depending on the nature of the task) spawn a new stack frame. That makes sense, as the OS/running task has to keep track of concurrent (or even parallel) things. Sharing one frame wouldn't do the trick.

The delegate part is interesting. So basically, one stack frame pointing to another stack frame :D 

I say my hey to `ThreadLocal`. Not sure how that things behaves (well that is why we have `AsyncLocal` for async/await)

**Reply by toupswork** [reply]
Not sure I caught the last part of your message.

**Reply by linkdotnet** [reply]
Sorry - shouldn’t stop and start typing between other things.

I wanted to amend that the stack frames per thread are also the root cause why ThreadLocal doesn’t work I async and await

**Reply by toupswork** [reply]
Ahh.. That makes total sense then.
